### ISSUE TITLE
[java] AvoidInstantiatingObjectsInLoops - false positive with Collections

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#3088]
Could you please reopen this? I still experience the false positive with 6.30.0, even when using an example adapted from your tests:

```
public Collection<Car> checkCollection() {
    Collection<Car> thisIsACar = new ArrayList<>();
    for(int i = 0; i < 3; ++i) {
        thisIsACar.add(new Car());
    }
    return thisIsACar;
}

private static class Car {
}
```

The array example did not cause a false positive warning with 6.29.0, and still does not with 6.30.0. For created objects as shown above both versions show the warning.

_Originally posted by @C-Otto in https://github.com/pmd/pmd/issues/2207#issuecomment-766723978_

**Comment by adangel** [comment]
See discussion at #2207 

* The bug can be reproduced with PMD 6.30.0
* The bug is fixed with the latest version 6.31.0 via #3061 
* Before #3061 the statement expression had only the result type of the method call. The information about the receiver type (in that case the collection `cars`) was lost and therefore we couldn't detect, that we were dealing with a collection. With that fix, the rule has also been adjusted to actually look at the receiver type instead of the result type (which simply was wrong before).
